File: 백준/Silver/14916. 거스름돈/거스름돈.java

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(br.readLine());

        ArrayList<Integer> arr = new ArrayList<>();
        int cnt = 0;

        int big = N / 5;
        for (int i = 0; i < big; i++) {
            arr.add(5);
            cnt += 5;
        }

        while (true) {
            int num = N - cnt;
            if ((num % 2) == 0) {
                for (int i = 0; i < (num / 2); i++) {
                    arr.add(2);
                    cnt += 2;
                }
                break;
            } else {
                if (arr.size() == 0) {
                    break;
                } else {
                    arr.remove(0);
                    cnt -= 5;
                }
            }
        }

        if (arr.size() == 0) {
            System.out.println(-1);
        } else {
            System.out.println(arr.size());
        }

    }
}

File: 백준/Silver/14916. 거스름돈/거스름돈.py
import sys

N = int(sys.stdin.readline())
list = []
cnt = 0

big = N // 5
for i in range(big):
    list.append(5)
    cnt += 5

while (True):
    num = (N - cnt) % 2
    if (num == 0):
        for i in range((N - cnt) // 2):
            list.append(2)
            cnt += 2
        break
    else:
        if (len(list) == 0):
            break
        else:
            list.pop()
            cnt -= 5


if len(list) == 0:
    print(-1)
else:
    print(len(list))
